Fix ORA-00933 error when executing `rails db:schema:load`

koic opened the pull request:
ORA-00933 error occurred when executing `rails db:schema:load`.

I think that it is probably due to not being able to process `/` by ruby-oci8. In this PR it is a proposal to insert multiple lines without using `/`.

I'm added a reproduction test in active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb.

This PR passed the following reproduction test.

```sh
Failures:

  1) OracleEnhancedAdapter schema definition load schema should loads the migration schema table from insert versions sql
       Failure/Error:
         expect {
           @conn.execute insert_versions_sql
         }.not_to raise_error

         expected no Exception, got #<ActiveRecord::StatementInvalid: OCIError: ORA-00933: SQL command not properly ended:
INSERT INTO sc...on) VALUES ('20160102000000')

        /

       INSERT INTO schema_migrations (version) VALUES ('20160103000000')> with backtrace:
         # stmt.c:243:in oci8lib_230.so
         # /home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/ruby-oci8-d3a1a3dedde6/lib/oci8/cursor.rb:129:in `exec'
         # /home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/ruby-oci8-d3a1a3dedde6/lib/oci8/oci8.rb:276:in `exec_internal'
         # /home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/ruby-oci8-d3a1a3dedde6/lib/oci8/oci8.rb:267:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:447:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:90:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `block in execute'
         #
/home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/rails-a19810583625/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:589:in `block in log'
         #
/home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/rails-a19810583625/activesupport/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
         #
/home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/rails-a19810583625/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:583:in `log'
         # ./lib/active_record/connection_adapters/oracle_enhanced_adapter.rb:1284:in `log'
         # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `execute'
         # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1449:in `block (4 levels) in <top
(required)>'
```

Related PR is #1074 .

Thanks.

yahonda commented:
cc @jlahtinen

jlahtinen commented:
Looks good and should be faster than before. I have no idea why / character is problem here. After this fix our structure file contains 599 / characters :). I also tested our migrations and loading from structure with this pull request and found no problems.

yahonda commented:
@jlahtinen Thanks for the answer. Good to know this pull request works in your environment.

yahonda commented:
Thanks for opening a pull request.

I'd like to know some end user examples how many records are stored in `schema_migrations` table.
Since `INSERT ALL` statement have some restrictions with Oracle database 11.1 or lower:

https://docs.oracle.com/cd/B19306_01/server.102/b14200/statements_9014.htm#i2080134
https://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_9014.htm#i2080134

```
Restrictions on Multitable Inserts Multitable inserts are subject to the following restrictions:
... snip ...

    In a multitable insert, all of the insert_into_clauses cannot combine to specify more than 999 target columns.
...
```
Note: This restriction does not apply to 11.2 and 12.1.

https://docs.oracle.com/cloud/latest/db112/SQLRF/statements_9014.htm#i2080134
https://docs.oracle.com/database/121/SQLRF/statements_9014.htm#i2080134

Let's say some users running Oracle database 10.2 have more than 1000 migrations history in `schema_migrations` table,  These 1000+ records need to be loaded via `rails db:schema:load`

@koic 
It is very helpful if you add some test cases to load 1000 migration history. 
of course, it is also helpful that "No customers using Oracle database 10.2 anymore" or "We do not have 1000 migrations" kind of input.

Thanks.

yahonda commented:
We may not need to add tests which actually creates 1000 schema_migrations history if it makes our rspec much slower. but at least I would like to know if it can support more than 1000 rows inserted at the same time using `insert all`.

koic commented:
Thanks for your reply and advice.

I use Oracle 11.2, so I can't see the restriction behavior in other versions of Oracle.

> I would like to know if it can support more than 1000 rows inserted at the same time using insert all.

I did try it. When I checked the operation of 1,000 migrations in Oracle 11.2 environment, it seems that there was no problem.

I used the following code for the test. This test code based on this PR codes.

https://github.com/koic/oracle-enhanced/blob/75ca620925c486a827a24f6bc1ce82febc3dc4cf/spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb#L1433-L1460

> if it makes our rspec much slower.

Yes. Actually, it seemed that there was not much value to leave as a test case, so I'm not putting it into this PR commitment.

yahonda commented:
Thanks for adding a new spec. Installed 10.2.0.5 database and found it reproduces `ORA-24335` as expected.

`OCIError: ORA-24335: cannot support more than 1000 columns: INSERT ...('20160101000999')`

Would you update your pull request like:

* [`supports_multi_insert?`](https://github.com/rsim/oracle-enhanced/blob/master/lib/active_record/connection_adapters/oracle_enhanced_adapter.rb#L495-L497) returns `true` when database version is 11.2 or higher. ( I do not have a 11.1 environment then let's follow the [11.1 documentation](https://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_9014.htm#i2080134)

* if `supports_multi_insert?` returns `true` `insert_version_sql` returns `INSERT ALL` , else it returns each insert statement per row. No need to think about separating insert all statement per 999 or something like that.



```sql
SQL> select * from v$version;

BANNER
----------------------------------------------------------------
Oracle Database 10g Enterprise Edition Release 10.2.0.5.0 - 64bi
PL/SQL Release 10.2.0.5.0 - Production
CORE	10.2.0.5.0	Production
TNS for Linux: Version 10.2.0.5.0 - Production
NLSRTL Version 10.2.0.5.0 - Production

SQL> 
```

```ruby
$ rspec ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1449
==> Loading config from ENV or use default
==> Running specs with MRI version 2.3.3
==> Effective ActiveRecord version 5.0.0.1
Run options: include {:locations=>{"./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb"=>[1449]}}
F

Failures:

  1) OracleEnhancedAdapter schema definition load schema should loads the migration schema table from insert versions sql
     Failure/Error:
       expect {
         @conn.execute insert_versions_sql
       }.not_to raise_error
     
       expected no Exception, got #<ActiveRecord::StatementInvalid: OCIError: ORA-24335: cannot support more than 1000 columns: INSERT ...('20160101000999')
       INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160101001000')
       SELECT * FROM DUAL
       > with backtrace:
         # stmt.c:243:in oci8lib_230.so
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/ruby-oci8-c2582a6c2441/lib/oci8/cursor.rb:129:in `exec'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/ruby-oci8-c2582a6c2441/lib/oci8/oci8.rb:276:in `exec_internal'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/ruby-oci8-c2582a6c2441/lib/oci8/oci8.rb:267:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:447:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:90:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `block in execute'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-4db7540caf1f/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:589:in `block in log'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-4db7540caf1f/activesupport/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-4db7540caf1f/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:583:in `log'
         # ./lib/active_record/connection_adapters/oracle_enhanced_adapter.rb:1284:in `log'
         # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `execute'
         # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1451:in `block (4 levels) in <top (required)>'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/matchers/built_in/raise_error.rb:52:in `matches?'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/matchers/built_in/raise_error.rb:71:in `does_not_match?'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:78:in `does_not_match?'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:72:in `block in handle_matcher'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:27:in `with_matcher'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:70:in `handle_matcher'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/expectation_target.rb:78:in `not_to'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/expectation_target.rb:106:in `not_to'
         # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1450:in `block (3 levels) in <top (required)>'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `instance_exec'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `block in run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `block in with_around_and_singleton_context_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `block in with_around_example_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `block in run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:602:in `run_around_example_hooks_for'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `with_around_example_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `with_around_and_singleton_context_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:251:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:627:in `block in run_examples'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `map'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `run_examples'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:589:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `block in run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `map'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (3 levels) in run_specs'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `map'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (2 levels) in run_specs'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/configuration.rb:1835:in `with_suite_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:112:in `block in run_specs'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/reporter.rb:77:in `report'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:111:in `run_specs'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:87:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:71:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:45:in `invoke'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/exe/rspec:4:in `<top (required)>'
         # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `load'
         # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `<main>'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-support-3.5.0/lib/rspec/support.rb:87:in `block in <module:Support>'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-support-3.5.0/lib/rspec/support.rb:96:in `notify_failure'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/fail_with.rb:27:in `fail_with'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:40:in `handle_failure'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:72:in `block in handle_matcher'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:27:in `with_matcher'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:70:in `handle_matcher'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/expectation_target.rb:78:in `not_to'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/expectation_target.rb:106:in `not_to'
     # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1450:in `block (3 levels) in <top (required)>'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `instance_exec'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `block in with_around_and_singleton_context_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `block in with_around_example_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:602:in `run_around_example_hooks_for'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `with_around_example_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `with_around_and_singleton_context_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:251:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:627:in `block in run_examples'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `run_examples'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:589:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (3 levels) in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (2 levels) in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/configuration.rb:1835:in `with_suite_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:112:in `block in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/reporter.rb:77:in `report'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:111:in `run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:87:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:71:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:45:in `invoke'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/exe/rspec:4:in `<top (required)>'
     # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `load'
     # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `<main>'

Finished in 2.65 seconds (files took 0.70116 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1449 # OracleEnhancedAdapter schema definition load schema should loads the migration schema table from insert versions sql

$
```

jlahtinen commented:
I feel obligated to say that imho `INSERT ALL` is not multi insert in a way rails codebase sees it.

koic commented:
@yahonda @jlahtinen Thanks for information.

> supports_multi_insert? returns true when database version is 11.2 or higher. I do not have a 11.1 environment then let's
follow the 11.1 documentation

It fixed by this commit. https://github.com/rsim/oracle-enhanced/pull/1084/commits/1fdef9fc61d54f796cf45e0cf49bebe3a8996eef

> if supports_multi_insert? returns true insert_version_sql returns INSERT ALL

It fixed by this commit. https://github.com/rsim/oracle-enhanced/pull/1084/commits/92698197695633d340e9dffcaa81cabc75c937cc

> else it returns each insert statement per row. No need to think about separating insert all statement per 999 or something
like that.

Here's a problem.

First, I'm using the implementation based on https://github.com/rsim/oracle-enhanced/pull/1074 . Because I can't be it returns each insert statement per row.

For example, executing each insert with "a semicolon" at the end of an insert statement results in an ORA-00911 error.

```
  1) OracleEnhancedAdapter schema definition load schema false should loads the migration schema table from insert versions sql
     Failure/Error:
       expect {
         @conn.execute insert_versions_sql
       }.not_to raise_error

       expected no Exception, got #<ActiveRecord::StatementInvalid: OCIError: ORA-00911: invalid character: INSERT INTO
"SCHEMA_MIGRATI...n) VALUES ('20160101000999');

       INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160101001000');> with backtrace:
         # stmt.c:243:in oci8lib_230.so
         # /home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/ruby-oci8-125488941e62/lib/oci8/cursor.rb:129:in `exec'
         # /home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/ruby-oci8-125488941e62/lib/oci8/oci8.rb:276:in `exec_internal'
         # /home/vagrant/.rvm/gems/ruby-2.3.3/bundler/gems/ruby-oci8-125488941e62/lib/oci8/oci8.rb:267:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:447:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:90:in `exec'
```

I think that the following commit is necessary to solve this problem. With this commit it will be able to run each insert.

https://github.com/rails/rails/commit/fd87169eb11fc4cfd9082dabe0a85f3bfa385c29

How about pending as a known problem until it can use above commit?

https://github.com/rsim/oracle-enhanced/pull/1084/commits/92698197695633d340e9dffcaa81cabc75c937cc#diff-88088c2f724c15cb81eee850463a1327R1467

That problem remain that can't process `/` by ruby-oci8. This is about when `supports_multi_insert?` is false.

jlahtinen commented:
@koic is there any way you could show us your structure.sql? Do you know the reasons why your structure.sql does not contain any `/` characters (or am I understood something wrong)?.

Our structure.sql has lots of stuff like

    CREATE SEQUENCE "FOO_SEQ" MINVALUE 1 MAXVALUE 9999999999999999999999999999 INCREMENT BY 1 ORDER NOCYCLE

    /

    CREATE TABLE "FOO" (
    "ID" NUMBER(38,0) NOT NULL,
     "NAME" VARCHAR2(100),
     "CREATED_AT" TIMESTAMP(6) NOT NULL,
     "UPDATED_AT" TIMESTAMP(6) NOT NULL,
     CONSTRAINT SYS_C001 PRIMARY KEY (ID)
    )

    /

    ALTER TABLE "BAR" ADD CONSTRAINT "BAR" FOREIGN KEY ("FOO_ID") REFERENCES "FOO"("ID")

    /

koic commented:
@jlahtinen Thanks for your reply.

My `db/structure.sql` also contains a lot of `/` too. But I use `rails db:schema:load`. It is not `rails db:structure:load`.

Therefore `db/schema.rb` is used instead of` db/structure.sql`.

In my seeing, `rails db:structure:load` executes the` execute` method separately by removing `/`. This is not a problem

https://github.com/rsim/oracle-enhanced/blob/v1.7.8/lib/active_record/connection_adapters/oracle_enhanced/structure_dump.rb#L291-L295

On the other hand, `rails db:schema:load` executes the `execute` method as it is without removing` / `.

https://github.com/rails/rails/blob/v5.0.0.1/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L1041

What I'm in trouble is `rails db:schema:load`. Thanks.

yahonda commented:
@koic Would you open a pull request or open an issue at rails/rails to backport https://github.com/rails/rails/commit/fd87169eb11fc4cfd9082dabe0a85f3bfa385c29 into `5-0-stable` branch with a summary discussed here.

yahonda commented:
@koic 

I'd like to know if backporting `fd87169eb11fc4cfd9082dabe0a85f3bfa385c29` into 5-0-stable branch is good idea to remove this pending spec.

#### Changes made

- Backport https://github.com/rails/rails/commit/fd87169eb11fc4cfd9082dabe0a85f3bfa385c29 into 5-0-stable branch called `5-0-stable_fd87169`
- Updated `fix_ora_00933_error` branch adding two commits to run pending spec and use `5-0-stable_fd87169` branch for testing

Still, this spec gets failed. Would like to know if I misunderstand something.

#### Environment:

- ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-linux]
- Oracle Database 12c Enterprise Edition Release 12.1.0.2.0 - 64bit Production

```ruby
$ rspec ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1464
==> Loading config from ENV or use default
==> Running specs with MRI version 2.3.3
==> Effective ActiveRecord version 5.0.0.1
Run options: include {:locations=>{"./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb"=>[1464]}}
F

Failures:

  1) OracleEnhancedAdapter schema definition load schema supports_multi_insert is false should loads the migration schema table from insert versions sql
     Failure/Error:
       expect {
         @conn.execute insert_versions_sql
       }.not_to raise_error

       expected no Exception, got #<ActiveRecord::StatementInvalid: OCIError: ORA-00933: SQL command not properly ended: INSERT INTO "S...) VALUES ('20160102000000')

       /

       INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160103000000')> with backtrace:
         # stmt.c:243:in oci8lib_230.so
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/ruby-oci8-c2582a6c2441/lib/oci8/cursor.rb:129:in `exec'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/ruby-oci8-c2582a6c2441/lib/oci8/oci8.rb:276:in `exec_internal'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/ruby-oci8-c2582a6c2441/lib/oci8/oci8.rb:267:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:447:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/oci_connection.rb:90:in `exec'
         # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `block in execute'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-4fe36f25f26c/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:589:in `block in log'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-4fe36f25f26c/activesupport/lib/active_support/notifications/instrumenter.rb:21:in `instrument'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-4fe36f25f26c/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:583:in `log'
         # ./lib/active_record/connection_adapters/oracle_enhanced_adapter.rb:1284:in `log'
         # ./lib/active_record/connection_adapters/oracle_enhanced/database_statements.rb:11:in `execute'
         # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1470:in `block (5 levels) in <top (required)>'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/matchers/built_in/raise_error.rb:52:in `matches?'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/matchers/built_in/raise_error.rb:71:in `does_not_match?'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:78:in `does_not_match?'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:72:in `block in handle_matcher'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:27:in `with_matcher'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:70:in `handle_matcher'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/expectation_target.rb:78:in `not_to'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/expectation_target.rb:106:in `not_to'
         # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1469:in `block (4 levels) in <top (required)>'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `instance_exec'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `block in run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `block in with_around_and_singleton_context_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `block in with_around_example_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `block in run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:602:in `run_around_example_hooks_for'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `with_around_example_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `with_around_and_singleton_context_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:251:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:627:in `block in run_examples'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `map'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `run_examples'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:589:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `block in run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `map'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `block in run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `map'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (3 levels) in run_specs'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `map'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (2 levels) in run_specs'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/configuration.rb:1835:in `with_suite_hooks'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:112:in `block in run_specs'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/reporter.rb:77:in `report'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:111:in `run_specs'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:87:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:71:in `run'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:45:in `invoke'
         # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/exe/rspec:4:in `<top (required)>'
         # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `load'
         # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `<main>'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-support-3.5.0/lib/rspec/support.rb:87:in `block in <module:Support>'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-support-3.5.0/lib/rspec/support.rb:96:in `notify_failure'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/fail_with.rb:27:in `fail_with'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:40:in `handle_failure'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:72:in `block in handle_matcher'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:27:in `with_matcher'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/handler.rb:70:in `handle_matcher'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/expectation_target.rb:78:in `not_to'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-expectations-3.5.0/lib/rspec/expectations/expectation_target.rb:106:in `not_to'
     # ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1469:in `block (4 levels) in <top (required)>'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `instance_exec'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `block in with_around_and_singleton_context_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `block in with_around_example_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:602:in `run_around_example_hooks_for'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `with_around_example_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `with_around_and_singleton_context_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:251:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:627:in `block in run_examples'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `run_examples'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:589:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (3 levels) in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (2 levels) in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/configuration.rb:1835:in `with_suite_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:112:in `block in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/reporter.rb:77:in `report'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:111:in `run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:87:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:71:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:45:in `invoke'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/exe/rspec:4:in `<top (required)>'
     # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `load'
     # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `<main>'

Finished in 0.22197 seconds (files took 1.47 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/active_record/connection_adapters/oracle_enhanced_schema_statements_spec.rb:1464 # OracleEnhancedAdapter schema definition load schema supports_multi_insert is false should loads the migration schema table from insert versions sql

$
```

ojiry commented:
I think that this PR doesn't assume backport. After the backport is merged, I think that it will be changed as follows.

```rb
if versions.is_a?(Array)
  versions.inject("INSERT ALL\n") { |sql, version|
    sql << "INTO #{quote_table_name(sm_table)} (version) VALUES (#{quote(version)})\n"
  } << "SELECT * FROM DUAL\n"
else
  "INSERT INTO #{quote_table_name(sm_table)} (version) VALUES (#{quote(versions)})"
end
```

For that reason the test should also be changed.

Thanks.

⚠️ This English is made with Google Translate. If I have said something rude, I'm sorry.

koic commented:
@yahonda It is my expected behavior. You're not wrong. And @ojiry explained it (Thanks for your follow-up) .

I think that again. I have 4 ideas.

- idea 1: Merge this PR as it is, then fix the implementation later. This idea is @ojiry explanatory approach. As the result, we get to resolve ORA-00933 error when Oracle 11.2+.
- idea 2: The basic is the same as idea 1. In addition, remove pending spec code until needed. As the result, If the presence of the pending spec code leads to confusion, we can resolve it.
- idea 3: I'll fix production code with @ojiry proposal implementation at this PR. However, we will not expected behavior until new AR released. As the result, that codes is probably what we expected.
- idea 4: Wait for https://github.com/rails/rails/commit/fd87169eb11fc4cfd9082dabe0a85f3bfa385c29 to be backported. As the result, We can make it completely expected code base. But, I don't know when will be released that includes https://github.com/rails/rails/commit/fd87169eb11fc4cfd9082dabe0a85f3bfa385c29 .

I think that 1, 2 or 3 is better.

yahonda commented:
Let's take option 3.

Would you update this pull request to make every spec works once https://github.com/rails/rails/commit/fd87169eb11fc4cfd9082dabe0a85f3bfa385c29 merged to 5-0-stable as requested in rails/rails#27379 . 

Thanks.

koic commented:
Sure, I updated code by commit below. And In that code, I'm rewriting a part of the spec code that I misunderstood.

https://github.com/rsim/oracle-enhanced/pull/1084/commits/42c8fcba35bb4303e2c75a1b870360b2295bda11

In addition, backporting PR https://github.com/rails/rails/pull/27379 has been merged.

This test specs will gets succeed with above codes.

Environments are:

* ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-linux]
* Oracle Database 11g Express Edition Release 11.2.0.2.0 - 64bit Production

Thanks.

yahonda commented:
Thanks for updating the pull request and providing useful information at rails/rails#27379 .

Would you take a look at these 2 items and update your pull requests. I understand you do not have 10gR2 environment then please let me know if you need to have 10gR2 environment to address the failure. I'll create a temporary one and give you an access.

- 4 specs are failing with JRuby

There are [4 specs failing](https://travis-ci.org/yahonda/oracle-enhanced/jobs/184929451) when tested with JRuby. They are almost like testing Array#to_s then
I think we are ok to remove these 4 specs. 

- One spec fails when tested with Oracle Database 10gR2 10.2.0.5

### Environment:

- ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-linux]
- Oracle Database 12c Enterprise Edition Release 12.1.0.2.0 - 64bit Production
- Tested with the latest `5-0-stable` branch 

```ruby
$ rspec ./spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb:302
==> Loading config from ENV or use default
==> Running specs with MRI version 2.3.3
==> Effective ActiveRecord version 5.0.0.1
Run options: include {:locations=>{"./spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb"=>[302]}}
F

Failures:

  1) OracleEnhancedAdapter structure dump schema migrations should dump schema migrations one version per insert
     Failure/Error: super
     
     TypeError:
       can't quote Array
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-48f3e702b968/activerecord/lib/active_record/connection_adapters/abstract/quoting.rb:177:in `_quote'
     # ./lib/active_record/connection_adapters/oracle_enhanced/quoting.rb:100:in `_quote'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-48f3e702b968/activerecord/lib/active_record/connection_adapters/abstract/quoting.rb:22:in `quote'
     # ./lib/active_record/connection_adapters/oracle_enhanced/quoting.rb:88:in `quote'
     # ./lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb:128:in `insert_versions_sql'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/bundler/gems/rails-48f3e702b968/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:989:in `dump_schema_information'
     # ./lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb:117:in `dump_schema_information'
     # ./spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb:301:in `block (3 levels) in <top (required)>'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/memoized_helpers.rb:295:in `block (2 levels) in let'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/memoized_helpers.rb:157:in `block (3 levels) in fetch_or_store'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/memoized_helpers.rb:157:in `fetch'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/memoized_helpers.rb:157:in `block (2 levels) in fetch_or_store'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-support-3.5.0/lib/rspec/support/reentrant_mutex.rb:23:in `synchronize'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/memoized_helpers.rb:156:in `block in fetch_or_store'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/memoized_helpers.rb:155:in `fetch'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/memoized_helpers.rb:155:in `fetch_or_store'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/memoized_helpers.rb:295:in `block in let'
     # ./spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb:303:in `block (3 levels) in <top (required)>'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `instance_exec'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:254:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `block in with_around_and_singleton_context_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `block in with_around_example_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:602:in `run_around_example_hooks_for'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/hooks.rb:464:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:453:in `with_around_example_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:496:in `with_around_and_singleton_context_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example.rb:251:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:627:in `block in run_examples'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:623:in `run_examples'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:589:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `block in run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/example_group.rb:590:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (3 levels) in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `map'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:113:in `block (2 levels) in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/configuration.rb:1835:in `with_suite_hooks'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:112:in `block in run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/reporter.rb:77:in `report'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:111:in `run_specs'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:87:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:71:in `run'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:45:in `invoke'
     # /home/yahonda/.rbenv/versions/2.3.3/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/exe/rspec:4:in `<top (required)>'
     # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `load'
     # /home/yahonda/.rbenv/versions/2.3.3/bin/rspec:22:in `<main>'

Finished in 0.2569 seconds (files took 0.67569 seconds to load)
1 example, 1 failure

Failed examples:

rspec ./spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb:302 # OracleEnhancedAdapter structure dump schema migrations should dump schema migrations one version per insert

$
```

koic commented:
@yahonda Thanks for your confirmation. That really helps.
I want to rent 10gR2 environment.

yahonda commented:
@koic Sounds good. Will send an email to your gmail account about 10gR2 environment.

koic commented:
Thanks for your kindness.

It was a failure spec dependent on the version of Oracle. Therefore, there was a problem that the result changed in the runtime environment.
I fixed that spec code with this commit https://github.com/rsim/oracle-enhanced/pull/1084/commits/4b35f8b7c214c186fb8015e53969127578f3ff54.

> They are almost like testing Array#to_s then I think we are ok to remove these 4 specs

I see. I fixed it too https://github.com/rsim/oracle-enhanced/pull/1084/commits/ce8cb9779dec5e48d37c688ca22250d0a31b7b6f.

yahonda commented:
Thanks for your work.

https://github.com/rsim/oracle-enhanced/pull/1084/commits/4b35f8b7c214c186fb8015e53969127578f3ff54

This commit resolves "TypeError: can't quote Array" itself when tested with Oracle Database 10gR2 and CRuby.
I'm little bit confused with this commit since it returns `supports_multi_insert?` true when Oracle Database 10gR2 is used, which should return `false`.


```ruby
$ git diff
diff --git a/lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb b/lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb
index db9e7c8..497a1a4 100644
--- a/lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb
+++ b/lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb
@@ -121,10 +121,12 @@ module ActiveRecord
           sm_table = ActiveRecord::Migrator.schema_migrations_table_name
 
           if self.supports_multi_insert?
+            puts "multi insert is supported"
             versions.inject("INSERT ALL\n") { |sql, version|
               sql << "INTO #{quote_table_name(sm_table)} (version) VALUES (#{quote(version)})\n"
             } << "SELECT * FROM DUAL\n"
           else
+            puts "multi insert is NOT supported"
             "INSERT INTO #{quote_table_name(sm_table)} (version) VALUES (#{quote(versions)})"
           end
         end
$
```

```
$ rspec ./spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb:304
==> Loading config from ENV or use default
==> Running specs with MRI version 2.3.3
==> Effective ActiveRecord version 5.0.0.1
Run options: include {:locations=>{"./spec/active_record/connection_adapters/oracle_enhanced_structure_dump_spec.rb"=>[304]}}
multi insert is supported
.

Finished in 0.17527 seconds (files took 1.14 seconds to load)
1 example, 0 failures
$ 
```

yahonda commented:
Let me summarize my understanding how this pull request will work. 


- Expected `supports_multi_insert?` return value per database version:

    return `false` when Oracle Database 11.1 or lower 
    return `true`  when Oracle Database 11.2 or higher

- Expected dump file format per `supports_multi_insert?` return values:

	When `true` one insert all into statement inserts all migrations.

	```
	INSERT ALL 
    INTO "SCHEMA_MIGRATIONS" (version) VALUES ('...')
    INTO "SCHEMA_MIGRATIONS" (version) VALUES ('...')
    INTO "SCHEMA_MIGRATIONS" (version) VALUES ('...')
    INTO "SCHEMA_MIGRATIONS" (version) VALUES ('...')
	SELECT * FROM DUAL
	```

	when `false` each insert statement per migration.

	```
	INSERT INTO "SCHAMA_MIGRATIONS" (version) VALUES ('...')
	INSERT INTO "SCHAMA_MIGRATIONS" (version) VALUES ('...')
	INSERT INTO "SCHAMA_MIGRATIONS" (version) VALUES ('...')
	INSERT INTO "SCHAMA_MIGRATIONS" (version) VALUES ('...')
	```
	(repeat insert into statement per migration)


Based on the expected specs, I think the "should dump schema migrations one version per insert"
should test only when the database version is 11.2 or higher.
If the database version is 11.1 or lower, this spec needs skipped.

- "OracleEnhancedAdapter structure dump schema migrations should dump schema migrations one version per insert"

```
    it "should dump schema migrations one version per insert" do
      expect(dump).to eq <<-SQL.strip_heredoc
        INSERT ALL
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160101000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160102000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160103000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160104000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160105000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160106000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160107000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160108000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160109000000')
        INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160110000000')
        SELECT * FROM DUAL
      SQL
    end
```

If database version is 11.1 or lower, we can test to see if the dump generates like this:

```
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160101000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160102000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160103000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160104000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160105000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160106000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160107000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160108000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160109000000')
    INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160110000000')
```

yahonda commented:
One more update. Oracle enhanced adapter specs should pass using both CRuby with OCI driver and JRuby with JDBC driver. 

Thanks again for your work. Happy holidays!

koic commented:
Oops! It's my mistake, you're right.
Thanks for your review. I'll try again.

Happy holidays to you, too!

koic commented:
I fixed my mistake with https://github.com/rsim/oracle-enhanced/pull/1084/commits/c5022ab03db4cd67f54bee7a0388d58733db1c67. And I confirmed it with the following matrix.

* Ruby ... MIR 2.3.3 / jruby-9.1.6.0
* Oracle ... 11g R2 / 10g R2

Please review it when you have time. Thanks.

koic commented:
> Based on the expected specs, I think the "should dump schema migrations one version per insert"
> should test only when the database version is 11.2 or higher.
> If the database version is 11.1 or lower, this spec needs skipped.

(snip)

> If database version is 11.1 or lower, we can test to see if the dump generates like this:

I'm sorry. I misunderstood about the above.
Try it again. Please give me some time.

koic commented:
I fixed the points that were pointed out.

> If database version is 11.1 or lower, we can test to see if the dump generates like this:

I have changed the implementation for this. It is the implementation based on #1074.

https://github.com/rsim/oracle-enhanced/pull/1084/files#diff-c995729a3e07c328a5502169ce79301eR128

> If database version is 11.1 or lower, we can test to see if the dump generates like this:

if database version is 11.1 or lower, I expect to be dumped as follows. And so implemented.

I think that expected SQL that should be able to run in Oracle 11.1 or lower (Need a `/`) .

```sql
INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160101000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160102000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160103000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160104000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160105000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160106000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160107000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160108000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160109000000')

/

INSERT INTO "SCHEMA_MIGRATIONS" (version) VALUES ('20160110000000')
```

I'm correcting it overall. It may be better to review it in [PR diff](https://github.com/rsim/oracle-enhanced/pull/1084/files).

yahonda commented on lib/active_record/connection_adapters/oracle_enhanced/schema_statements.rb:
```diff
@@ -118,11 +118,23 @@ def dump_schema_information #:nodoc:
-          versions.map { |version|
-            "INSERT INTO #{sm_table} (version) VALUES ('#{version}')"
-          }.join "\n\n/\n\n"
+          if self.supports_multi_insert?
```
Would you tell me if there is a reason it uses `self`. If not would you remove it.

koic replied:
I was writing `self` as a habit. And there is no problem remove it for consistency of coding style. I think that is better. I'll remove it.

yahonda commented:
Great job. I have one (probably last) question for this pull request about `self`.

koic commented:
Thanks for your careful review. I fixed the question that you pointed out.

yahonda commented:
Thanks. Please squash your commits.

koic commented:
Yes, I’m done. Thanks.

yahonda commented:
Merged. Thanks again for your great work.

koic commented:
Thanks for your review and merging.
These codes got quite better. I really appreciate it, many thanks!
